File: crates/juicefront/ui/src/lib/hosts.ts
/**
 * Curated storage-node lists for the host selector modal.
 *
 * Official nodes are run by the Juicebox team; unofficial (community) nodes
 * are run by volunteers. Both lists can be extended without a code change:
 *
 *   PUBLIC_OFFICIAL_HOSTS="Box EU|https://box.juicey.dev|EU,..."
 *   PUBLIC_UNOFFICIAL_HOSTS="Alice|https://alice.example.com|US,..."
 *
 * Entries are `Name|https://url|Region`; region is optional. Invalid entries
 * are ignored. Imported by `.astro` frontmatter only (never shipped as-is;
 * the component embeds the resolved lists as data attributes).
 */

export interface HostNode {
  name: string;
  url: string;
  region: string;
  official: boolean;
}

const OFFICIAL_DEFAULTS: HostNode[] = [
  { name: "Box", url: "https://box.juicey.dev", region: "", official: true },
  { name: "F", url: "https://f.juicey.dev", region: "", official: true },
];

const UNOFFICIAL_DEFAULTS: HostNode[] = [];

function normalizeUrl(raw: string): string | null {
  let h = raw.trim();
  if (!h) return null;
  if (!/^[a-zA-Z][a-zA-Z0-9+.-]*:\/\//.test(h)) h = `https://${h}`;
  try {
    const url = new URL(h);
    if (url.protocol !== "http:" && url.protocol !== "https:") return null;
    return url.origin;
  } catch {
    return null;
  }
}

function parseEnvList(
  raw: string | undefined,
  official: boolean,
): HostNode[] {
  if (!raw) return [];
  const out: HostNode[] = [];
  for (const entry of raw.split(",")) {
    const [name = "", url = "", region = ""] = entry.split("|").map((s) => s.trim());
    const base = normalizeUrl(url);
    if (!base) continue;
    out.push({ name: name || base, url: base, region, official });
  }
  return out;
}

function mergeNodes(base: HostNode[], extra: HostNode[]): HostNode[] {
  const seen = new Set(base.map((n) => n.url));
  const out = [...base];
  for (const n of extra) {
    if (seen.has(n.url)) continue;
    seen.add(n.url);
    out.push(n);
  }
  return out;
}

/** SSR-safe: reads env at render time (works in Astro frontmatter). */
export function getOfficialNodes(): HostNode[] {
  const env =
    typeof process !== "undefined"
      ? process.env.JUICEFRONT_OFFICIAL_HOSTS
      : undefined;
  const clientEnv =
    (import.meta.env.PUBLIC_OFFICIAL_HOSTS as string | undefined) ?? env;
  return mergeNodes(OFFICIAL_DEFAULTS, parseEnvList(clientEnv, true));
}

/** SSR-safe: reads env at render time (works in Astro frontmatter). */
export function getUnofficialNodes(): HostNode[] {
  const env =
    typeof process !== "undefined"
      ? process.env.JUICEFRONT_UNOFFICIAL_HOSTS
      : undefined;
  const clientEnv =
    (import.meta.env.PUBLIC_UNOFFICIAL_HOSTS as string | undefined) ?? env;
  return mergeNodes(UNOFFICIAL_DEFAULTS, parseEnvList(clientEnv, false));
}

File: crates/juicefront/src/ssr/hosts.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HostNode {
    pub name: String,
    pub url: String,
    pub region: String,
    pub official: bool,
}

const OFFICIAL_DEFAULTS: &[(&str, &str)] = &[
    ("Box", "https://box.juicey.dev"),
    ("F", "https://f.juicey.dev"),
];

fn normalize_url(raw: &str) -> Option<String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    let with_scheme = if trimmed.contains("://") {
        trimmed.to_owned()
    } else {
        format!("https://{trimmed}")
    };
    let parsed: url::Url = with_scheme.parse().ok()?;
    if parsed.scheme() != "http" && parsed.scheme() != "https" {
        return None;
    }
    Some(format!("{}://{}", parsed.scheme(), parsed.host_str()?))
}

fn parse_env_list(raw: Option<String>, official: bool) -> Vec<HostNode> {
    let Some(raw) = raw else {
        return Vec::new();
    };
    let mut out = Vec::new();
    for entry in raw.split(',') {
        let mut parts = entry.split('|').map(str::trim);
        let name = parts.next().unwrap_or("");
        let url = parts.next().unwrap_or("");
        let region = parts.next().unwrap_or("").to_owned();
        let Some(base) = normalize_url(url) else {
            continue;
        };
        out.push(HostNode {
            name: if name.is_empty() {
                base.clone()
            } else {
                name.to_owned()
            },
            url: base,
            region,
            official,
        });
    }
    out
}

fn merge_nodes(base: Vec<HostNode>, extra: Vec<HostNode>) -> Vec<HostNode> {
    let mut seen: Vec<String> = base.iter().map(|node| node.url.clone()).collect();
    let mut out = base;
    for node in extra {
        if seen.iter().any(|url| url == &node.url) {
            continue;
        }
        seen.push(node.url.clone());
        out.push(node);
    }
    out
}

pub fn official_nodes() -> Vec<HostNode> {
    let base: Vec<HostNode> = OFFICIAL_DEFAULTS
        .iter()
        .map(|(name, url)| HostNode {
            name: (*name).to_owned(),
            url: (*url).to_owned(),
            region: String::new(),
            official: true,
        })
        .collect();
    let extra = juiceutils::config::optional_secret("PUBLIC_OFFICIAL_HOSTS")
        .or_else(|| juiceutils::config::optional_secret("JUICEFRONT_OFFICIAL_HOSTS"));
    merge_nodes(base, parse_env_list(extra, true))
}

pub fn unofficial_nodes() -> Vec<HostNode> {
    let extra = juiceutils::config::optional_secret("PUBLIC_UNOFFICIAL_HOSTS")
        .or_else(|| juiceutils::config::optional_secret("JUICEFRONT_UNOFFICIAL_HOSTS"));
    merge_nodes(Vec::new(), parse_env_list(extra, false))
}
